portfolio.py

```diff
@@ -2,21 +2,47 @@
 Allocation of the asset.
 """
 
+import numpy as np
+from thousandaire.constants import TRADING_INSTRUMENTS
+
 class Portfolio(dict):
     """
     Set the portfolio
     """
-    def __init__(self, date):
+    def __init__(self, np_array=None, target=None):
         dict.__init__(self)
-        self.date = date
+        if np_array is not None:
+            if isinstance(np_array, np.ndarray):
+                self.decode_nparray(target, np_array)
+            else:
+                raise TypeError(
+                    'The input is %s, not a numpy array.' % type(np_array))
+
+    def decode_nparray(self, target, np_array):
+        """
+        Decode numpy array into portfolio.
+        """
+        if target is None:
+            raise KeyError('Decoded target is not found.')
```

Comment:
Finally, I though the judgment for target may put here in decoding process, since it a part of decoding.

Reply:
This makes sense.
However, the error message is very confusing, and it is possible that `target` is not in `TRADING_INSTRUMENTS` at all.
Can you change it to `'Invalid encoding method: %r" % target'`?